File: languages/go/monkey/parser/parser.go
package parser

import (
	"fmt"
	"monkey/ast"
	"monkey/lexer"
	"monkey/token"
	"strconv"
)

const (
	_ int = iota
	LOWEST
	EQUALS       // ==
	LESS_GREATER // > or <
	SUM          // +
	PRODUCT      // *
	PREFIX       // -X or !X
	CALL         // myFunction(X)
)

type (
	prefixParseFn func() ast.Expression               // doesn't have a left side
	infixParseFn  func(ast.Expression) ast.Expression // excepts the left side of the infix operator
)

type Parser struct {
	lex       *lexer.Lexer
	curToken  token.Token
	peekToken token.Token
	errors    []string

	prefixParseFns map[token.TokenType]prefixParseFn
	infixParseFns  map[token.TokenType]infixParseFn
}

func New(lex *lexer.Lexer) *Parser {
	p := &Parser{
		lex:    lex,
		errors: []string{},
	}

	p.prefixParseFns = make(map[token.TokenType]prefixParseFn)

	// We register a parsing function. If we encounter a token of
	// certain type, the it will call a mapped function
	p.registerPrefix(token.IDENT, p.parseIdentifier)
	p.registerPrefix(token.INT, p.parseIntegerLiteral)

	// Read two tokens, so curToken and peekToken are set
	p.nextToken()
	p.nextToken()

	return p
}

func (p *Parser) Errors() []string {
	return p.errors
}

func (p *Parser) ParseProgram() *ast.Program {
	program := &ast.Program{}
	program.Statements = []ast.Statement{}

	// put statements in program.Statements slice
	for !p.curTokenIs(token.EOF) {
		stmt := p.parseStatement()

		if stmt != nil {
			program.Statements = append(program.Statements, stmt)
		}

		p.nextToken()
	}

	return program
}

func (p *Parser) parseIdentifier() ast.Expression {
	return &ast.Identifier{
		Token: p.curToken,
		Value: p.curToken.Literal,
	}
}

func (p *Parser) parseIntegerLiteral() ast.Expression {
	intLiteral := &ast.IntegerLiteral{Token: p.curToken}

	value, err := strconv.ParseInt(p.curToken.Literal, 0, 64)

	if err != nil {
		msg := fmt.Sprintf("Could not parse %q as integer", p.curToken.Literal)
		p.errors = append(p.errors, msg)
		return nil
	}

	intLiteral.Value = value

	return intLiteral
}

func (p *Parser) peekError(tok token.TokenType) {
	msg := fmt.Sprintf("expected next token to be %s, got %s instead", tok, p.peekToken.Type)
	p.errors = append(p.errors, msg)
}

func (p *Parser) registerPrefix(tokenType token.TokenType, fn prefixParseFn) {
	p.prefixParseFns[tokenType] = fn
}

func (p *Parser) registerInfix(tokenType token.TokenType, fn infixParseFn) {
	p.infixParseFns[tokenType] = fn
}

func (p *Parser) parseStatement() ast.Statement {
	switch p.curToken.Type {
	case token.LET:
		return p.parseLetStatement()
	case token.RETURN:
		return p.parseReturnStatement()
	default:
		return p.parseExpressionStatement()
	}
}

func (p *Parser) nextToken() {
	p.curToken = p.peekToken
	p.peekToken = p.lex.NextToken()
}

// return <expression>
func (p *Parser) parseReturnStatement() *ast.ReturnStatement {
	stmt := &ast.ReturnStatement{Token: p.curToken}

	p.nextToken()

	// todo: skip here
	for !p.curTokenIs(token.SEMI) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseExpressionStatement() *ast.ExpressionStatement {
	stmt := &ast.ExpressionStatement{Token: p.curToken}
	stmt.Expression = p.parseExpression(LOWEST)

	if p.peekTokenIs(token.SEMI) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseExpression(precedence int) ast.Expression {
	prefix := p.prefixParseFns[p.curToken.Type]

	if prefix == nil {
		return nil
	}

	leftExp := prefix()

	return leftExp
}

// let <identifier> = <statement>
func (p *Parser) parseLetStatement() *ast.LetStatement {
	stmt := &ast.LetStatement{Token: p.curToken}

	if !p.expectPeek(token.IDENT) {
		return nil
	}

	stmt.Name = &ast.Identifier{
		Token: p.curToken,
		Value: p.curToken.Literal,
	}

	if !p.expectPeek(token.ASSIGN) {
		return nil
	}

	// todo: skip here
	for !p.curTokenIs(token.SEMI) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) curTokenIs(tok token.TokenType) bool {
	return p.curToken.Type == tok
}

func (p *Parser) expectPeek(tok token.TokenType) bool {
	if p.peekTokenIs(tok) {
		p.nextToken()
		return true
	}

	p.peekError(tok)

	return false
}

func (p *Parser) peekTokenIs(tok token.TokenType) bool {
	return p.peekToken.Type == tok
}

File: languages/php/liner/src/Parser/Parser.php
<?php

declare(strict_types=1);

namespace Serhii\Liner\Parser;

use Closure;
use Serhii\Liner\Ast\Expression;
use Serhii\Liner\Ast\ExpressionStatement;
use Serhii\Liner\Ast\Identifier;
use Serhii\Liner\Ast\IntegerLiteral;
use Serhii\Liner\Ast\StringLiteral;
use Serhii\Liner\Ast\Program;
use Serhii\Liner\Ast\ReturnStatement;
use Serhii\Liner\Ast\PutStatement;
use Serhii\Liner\Ast\Statement;
use Serhii\Liner\Lexer\Lexer;
use Serhii\Liner\Token\Token;
use Serhii\Liner\Token\TokenType;

class Parser
{
    private ?Token $curToken = null;
    private ?Token $peekToken = null;
    private array $errors = [];

    /**
     * @var array<string, Closure>
     */
    private array $prefixParseFns = [];

    /**
     * @var array<string, Closure>
     */
    private array $infixParseFns = [];

    public function __construct(private Lexer $lexer)
    {
        $this->nextToken();
        $this->nextToken();

        $this->registerPrefix(TokenType::INT, fn() => $this->parseIntegerLiteral());
        $this->registerPrefix(TokenType::IDENT, fn() => $this->parseIdentifier());
        $this->registerPrefix(TokenType::STR, fn() => $this->parseStringLiteral());
    }

    public function parseProgram(): Program
    {
        $statements = [];

        while (!$this->curTokenIs(TokenType::EOF)) {
            $stmt = $this->parseStatement();

            if ($stmt !== null) {
                $statements[] = $stmt;
            }

            $this->nextToken();
        }

        return new Program($statements);
    }

    public function errors(): array
    {
        return $this->errors;
    }

    private function registerPrefix(TokenType $tokenType, Closure $fn)
    {
        $this->prefixParseFns[$tokenType->value] = $fn;
    }

    private function registerInfix(TokenType $tokenType, Closure $fn)
    {
        $this->infixParseFns[$tokenType->value] = $fn;
    }

    private function curTokenIs(TokenType $tokenType): bool
    {
        return $this->curToken->type === $tokenType;
    }

    private function peekTokenIs(TokenType $tokenType): bool
    {
        return $this->peekToken->type === $tokenType;
    }

    private function nextToken(): void
    {
        $this->curToken = $this->peekToken;
        $this->peekToken = $this->lexer->nextToken();
    }

    private function parseStatement(): ?Statement
    {
        return match($this->curToken->type) {
            TokenType::RETURN => $this->parseReturnStatement(),
            TokenType::PUT => $this->parsePutStatement(),
            default => $this->parseExpressionStatement(),
        };
    }

    private function parseReturnStatement(): ?ReturnStatement
    {
        $token = $this->curToken;

        $this->nextToken();

        $stmt = new ReturnStatement($token, $this->parseExpression());

        if ($this->peekTokenIs(TokenType::PERIOD)) {
            $this->nextToken();
        }

        return $stmt;
    }

    private function parsePutStatement(): ?PutStatement
    {
        $token = $this->curToken;

        $this->nextToken();

        $exp = $this->parseExpression();

        if (!$this->expectPeek(TokenType::IN)) {
            return null;
        }

        $this->nextToken();

        $ident = $this->parseIdentifier();

        $stmt = new PutStatement($token, $ident, $exp);

        if ($this->peekTokenIs(TokenType::PERIOD)) {
            $this->nextToken();
        }

        return $stmt;
    }

    private function parseExpressionStatement(): ExpressionStatement
    {
        $stmt = new ExpressionStatement($this->curToken, $this->parseExpression());

        if ($this->peekTokenIs(TokenType::PERIOD)) {
            $this->nextToken();
        }

        return $stmt;
    }

    private function parseExpression(): ?Expression
    {
        $prefix = $this->prefixParseFns[$this->curToken->type->value] ?? null;

        if ($prefix === null) {
            $this->errors[] = sprintf(
                'no prefix parse function for token "%s" found',
                $this->curToken->type->value,
            );

            return null;
        }

        $leftExp = $prefix();

        while (!$this->peekTokenIs(TokenType::PERIOD)) {
            $infix = $this->infixParseFns[$this->peekToken->type->value] ?? null;

            if ($infix === null) {
                return $leftExp;
            }

            $this->nextToken();

            $leftExp = $infix($leftExp);
        }

        return $leftExp;
    }

    private function parseIntegerLiteral(): ?Expression
    {
        $num = (int) $this->curToken->literal;

        if (!is_numeric($this->curToken->literal)) {
            $this->errors[] = sprintf('"%s" is not numeric', $this->curToken->literal);
            return null;
        }

        return new IntegerLiteral($this->curToken, $num);
    }

    private function parseStringLiteral(): ?Expression
    {
        return new StringLiteral($this->curToken, $this->curToken->literal);
    }

    private function parseIdentifier(): ?Expression
    {
        return new Identifier($this->curToken, $this->curToken->literal);
    }

    private function expectPeek(TokenType $token): bool
    {
        if ($this->peekTokenIs($token)) {
            $this->nextToken();
            return true;
        }

        $this->errors[] = sprintf(
            'expected next token to be %s, got %s instead',
            $token->value,
            $this->peekToken->type->value,
        );

        return false;
    }
}